store-implementation/map-store/src/main/java/uk/gov/gchq/gaffer/mapstore/impl/MapImpl.java

```diff
@@ -15,61 +15,248 @@
+import uk.gov.gchq.gaffer.data.element.GroupedProperties;
 import uk.gov.gchq.gaffer.data.element.id.EdgeId;
 import uk.gov.gchq.gaffer.data.element.id.EntityId;
 import uk.gov.gchq.gaffer.mapstore.MapStoreProperties;
-import uk.gov.gchq.gaffer.store.StoreException;
+import uk.gov.gchq.gaffer.mapstore.factory.MapFactory;
+import uk.gov.gchq.gaffer.mapstore.factory.SimpleMapFactory;
+import uk.gov.gchq.gaffer.mapstore.multimap.MultiMap;
 import uk.gov.gchq.gaffer.store.schema.Schema;
 import uk.gov.gchq.gaffer.store.schema.SchemaElementDefinition;
+import uk.gov.gchq.gaffer.store.util.AggregatorUtil;
+import java.util.Collection;
+import java.util.Collections;
 import java.util.HashMap;
 import java.util.HashSet;
 import java.util.List;
 import java.util.Map;
 import java.util.Set;
+import java.util.stream.Stream;
 
 /**
  * The internal variables of this class are package-private. This allows operation handlers for the
  * {@link uk.gov.gchq.gaffer.mapstore.MapStore} to be placed in the same package and get access to the maps, without
  * exposing the internal state of the MapStore to classes outside of this package.
  */
 public class MapImpl {
+    public static final String AGG_ELEMENTS = "aggElements";
+    public static final String NON_AGG_ELEMENTS = "nonAggElements";
+    public static final String ENTITY_ID_TO_ELEMENTS = "entityIdToElements";
+    public static final String EDGE_ID_TO_ELEMENTS = "edgeIdToElements";
+    public static final Set<String> MAP_NAMES = Collections.unmodifiableSet(
+            Sets.newHashSet(AGG_ELEMENTS, NON_AGG_ELEMENTS));
+    public static final Set<String> MULTI_MAP_NAMES = Collections.unmodifiableSet(
+            Sets.newHashSet(ENTITY_ID_TO_ELEMENTS, EDGE_ID_TO_ELEMENTS));
 
-    static final String COUNT = "COUNT";
-
-    // elementToProperties maps from an Element containing the group-by properties to a Properties object without the
-    // group-by properties
-    Map<Element, Properties> elementToProperties;
-    // entityIdToElements is a map from an EntityId to the element key from elementToProperties
-    Map<EntityId, Set<Element>> entityIdToElements;
-    // edgeIdToElements is a map from an EdgeId to the element key from elementToProperties
-    Map<EdgeId, Set<Element>> edgeIdToElements;
-    final boolean maintainIndex;
-    final Map<String, Set<String>> groupToGroupByProperties = new HashMap<>();
-    final Map<String, Set<String>> groupToNonGroupByProperties = new HashMap<>();
-    final Set<String> groupsWithNoAggregation = new HashSet<>();
-    final Schema schema;
-    final List<String> aggregatedGroups;
-
-    public MapImpl(final Schema schema, final MapStoreProperties mapStoreProperties) throws StoreException {
+    /**
+     * aggElements maps from an Element containing the group-by properties
+     * to a Properties object without the group-by properties
+     */
+    private final Map<String, Map<Element, GroupedProperties>> aggElements = new HashMap<>();
```

Comment:
The original design tried to minimise the number of map look-ups that needed to be made before an element could be aggregated into the existing data. The element in the `Map<Element, Properties> elementToProperties` contained the group so only 1 look-up was needed. Now 2 are needed - 1 to get the right map for the group, and another to get the properties for the element. As the first look-up is always a local one (in a Java HashMap) the performance probably isn't worth worrying about, but is there any advantage in having the group as a key?

The Javadoc above this line needs updating as well.